src/Symfony/Component/Uri/Uri.php

```diff
@@ -0,0 +1,172 @@
+     * Uri::resolve('http://example.com/foo', '/bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo', 'bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo/', 'bar'); // http://example.com/foo/bar
+     * Uri::resolve('http://example.com/foo/', '../bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', '../../bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', '/bar'); // http://example.com/bar
+     * Uri::resolve('http://example.com/foo/', 'http://example.org/bar'); // http://example.org/bar
+     */
+    public static function resolve(self|string $baseUri, string $relativeUri): string
+    {
+        if ('' === $relativeUri) {
+            return (string) $baseUri;
+        }
+
+        // the relative URI is an absolute URI
+        if (str_contains($relativeUri, '://')) {
+            return $relativeUri;
+        }
+
+        $baseUri = $baseUri instanceof self ? $baseUri : self::parse($baseUri);
+        $relativeParts = explode('/', $relativeUri);
+        $baseParts = $baseUri->path && !str_starts_with($relativeUri, '/') ?
+            explode('/', trim($baseUri->path, '/'))
+            : [];
+
+        $resolvedPathSegments = $baseParts;
+        foreach ($relativeParts as $segment) {
+            if ('..' === $segment) {
+                array_pop($resolvedPathSegments);
+            } elseif ('.' !== $segment && '' !== $segment) {
+                $resolvedPathSegments[] = $segment;
+            }
+        }
+
+        $finalUri = clone $baseUri;
+        $finalUri->path = '/'.implode('/', $resolvedPathSegments);
+
+        return (string) $finalUri;
+    }
+
+    /**
+     * Returns a new instance with a new fragment text directive.
+     */
+    public function withFragmentTextDirective(string $start, ?string $end = null, ?string $prefix = null, ?string $suffix = null): static
+    {
+        $uri = clone $this;
+        $uri->fragmentTextDirective = new FragmentTextDirective($start, $end, $prefix, $suffix);
+
+        return $uri;
+    }
+
+    /**
+     * Returns a new instance with the host part of the URI converted to ASCII.
+     *
+     * @see https://www.unicode.org/reports/tr46/#ToASCII
+     */
+    public function withIdnHostAsAscii(): static
+    {
+        $uri = clone $this;
+        $uri->host = idn_to_ascii($uri->host, \IDNA_NONTRANSITIONAL_TO_ASCII, \INTL_IDNA_VARIANT_UTS46);
+
+        return $uri;
+    }
+
+    /**
+     * Returns a new instance with the host part of the URI converted to Unicode.
+     *
+     * @see https://www.unicode.org/reports/tr46/#ToUnicode
+     */
+    public function withIdnHostAsUnicode(): static
+    {
+        $uri = clone $this;
+        $uri->host = idn_to_utf8($uri->host, \IDNA_NONTRANSITIONAL_TO_UNICODE, \INTL_IDNA_VARIANT_UTS46);
+
+        return $uri;
+    }
+
+    public function __toString()
+    {
+        return $this->scheme.'://'
```

Comment:
`://` is not the correct delimiter, you should avoid using it anywhere in the cod, it will lead to errors down the line. If you want to strictly follow RFC3986 then `:` is to be added only if a scheme is present; likewise `//` must only be present when an authority is present. This representation will not work with URIs which do not have an authority part.

Reply:
Understood, thank you for the explanation! It's fixed now. Let's see if those oddities can be fixed here while we state something about using an external package 👍